File: backend/app/schemas/experimental_validation.py
"""STAMP Platform — Experimental Validation Schemas (v0.11-P1).

Pydantic models for wet-lab experimental validation runs and measurements.
All schemas enforce scientific-integrity constraints: no fabricated data,
no auto-promotion to EXPERIMENTALLY_VALIDATED.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.schemas.project import SchemaBase


# ---------------------------------------------------------------------------
# Enums as Literal strings (following existing codebase convention)
# ---------------------------------------------------------------------------

EXPERIMENT_TYPE_VALUES = [
    "MIC", "MBC", "HEMOLYSIS", "CYTOTOXICITY",
    "SERUM_STABILITY", "PROTEASE_STABILITY", "SALT_STABILITY",
    "BIOFILM", "RESISTANCE_INDUCTION", "OTHER",
]

RUN_STATUS_VALUES = ["PLANNED", "RUNNING", "COMPLETED", "FAILED", "INVALIDATED"]

VALIDATION_STATUS_VALUES = [
    "NOT_EXPERIMENTALLY_VALIDATED",
    "EXPERIMENT_PLANNED",
    "PARTIALLY_VALIDATED",
    "EXPERIMENTALLY_VALIDATED",
    "VALIDATION_FAILED",
]

METRIC_NAME_VALUES = [
    "MIC_ug_ml", "MBC_ug_ml", "hemolysis_percent",
    "HC50_ug_ml", "IC50_ug_ml", "cell_viability_percent",
    "serum_half_life_min", "protease_remaining_percent",
    "biofilm_inhibition_percent",
]

QUALITY_FLAG_VALUES = ["PASS", "WARNING", "FAILED", "NEEDS_REVIEW"]


# ---------------------------------------------------------------------------
# Base
# ---------------------------------------------------------------------------

class ExperimentalValidationRunBase(SchemaBase):
    """Base fields for an experimental validation run."""

    project_id: str
    candidate_id: str
    experiment_type: str = Field(..., max_length=50)
    organism: Optional[str] = Field(None, max_length=100)
    strain: Optional[str] = Field(None, max_length=100)
    protocol_name: Optional[str] = Field(None, max_length=255)
    protocol_version: Optional[str] = Field(None, max_length=50)
    operator: Optional[str] = Field(None, max_length=100)
    experiment_date: Optional[datetime] = None
    status: str = Field(default="PLANNED", max_length=20)
    notes: Optional[str] = None

    @field_validator("experiment_type")
    @classmethod
    def _validate_experiment_type(cls, v: str) -> str:
        if v not in EXPERIMENT_TYPE_VALUES:
            raise ValueError(f"experiment_type must be one of {EXPERIMENT_TYPE_VALUES}")
        return v

    @field_validator("status")
    @classmethod
    def _validate_status(cls, v: str) -> str:
        if v not in RUN_STATUS_VALUES:
            raise ValueError(f"status must be one of {RUN_STATUS_VALUES}")
        return v


class ExperimentalMeasurementBase(SchemaBase):
    """Base fields for an experimental measurement."""

    validation_run_id: str
    candidate_id: str
    metric_name: str = Field(..., max_length=50)
    value: Optional[float] = None
    unit: Optional[str] = Field(None, max_length=20)
    condition_json: dict[str, Any] = Field(default_factory=dict)
    replicate_id: Optional[str] = Field(None, max_length=20)
    raw_data_path: Optional[str] = Field(None, max_length=500)
    quality_flag: str = Field(default="PASS", max_length=20)

    @field_validator("metric_name")
    @classmethod
    def _validate_metric_name(cls, v: str) -> str:
        if v not in METRIC_NAME_VALUES:
            raise ValueError(f"metric_name must be one of {METRIC_NAME_VALUES}")
        return v

    @field_validator("quality_flag")
    @classmethod
    def _validate_quality_flag(cls, v: str) -> str:
        if v not in QUALITY_FLAG_VALUES:
            raise ValueError(f"quality_flag must be one of {QUALITY_FLAG_VALUES}")
        return v

    @field_validator("value")
    @classmethod
    def _validate_value(cls, v: Optional[float], info) -> Optional[float]:
        if v is None:
            return v
        data = info.data
        metric = data.get("metric_name", "")
        if metric in ("MIC_ug_ml", "MBC_ug_ml") and v < 0:
            raise ValueError(f"{metric} must be >= 0")
        if metric in ("hemolysis_percent", "cell_viability_percent", "protease_remaining_percent", "biofilm_inhibition_percent") and not (0 <= v <= 100):
            raise ValueError(f"{metric} must be between 0 and 100")
        return v


# ---------------------------------------------------------------------------
# Create / Update / Response
# ---------------------------------------------------------------------------

class ExperimentalValidationRunCreate(ExperimentalValidationRunBase):
    """Schema for creating a new validation run."""

    pass


class ExperimentalValidationRunUpdate(BaseModel):
    """Schema for updating a validation run (partial)."""

    model_config = ConfigDict(from_attributes=True)

    status: Optional[str] = Field(None, max_length=20)
    validation_status: Optional[str] = Field(None, max_length=50)
    notes: Optional[str] = None
    experiment_date: Optional[datetime] = None
    operator: Optional[str] = Field(None, max_length=100)

    @field_validator("status")
    @classmethod
    def _validate_status(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and v not in RUN_STATUS_VALUES:
            raise ValueError(f"status must be one of {RUN_STATUS_VALUES}")
        return v


class ExperimentalValidationRunResponse(ExperimentalValidationRunBase):
    """Schema for returning a validation run."""

    id: str
    validation_status: str = Field(default="NOT_EXPERIMENTALLY_VALIDATED", max_length=50)
    created_at: datetime
    updated_at: datetime


class ExperimentalMeasurementCreate(ExperimentalMeasurementBase):
    """Schema for creating a new measurement."""

    pass


class ExperimentalMeasurementResponse(ExperimentalMeasurementBase):
    """Schema for returning a measurement."""

    id: str
    created_at: datetime
    updated_at: datetime


# ---------------------------------------------------------------------------
# Aggregated / Summary schemas
# ---------------------------------------------------------------------------

class ExperimentalValidationSummaryResponse(BaseModel):
    """Summary of all experimental validation data for a candidate."""

    model_config = ConfigDict(from_attributes=True)

    candidate_id: str
    overall_validation_status: str
    run_count: int
    completed_run_count: int
    measurements: list[ExperimentalMeasurementResponse]


class CandidateExperimentalPriorityResponse(BaseModel):
    """Candidate priority response including experimental data."""

    model_config = ConfigDict(from_attributes=True)

    candidate_id: str
    composite_score: Optional[float] = None
    experimental_priority_score: Optional[float] = None
    priority_status: str
    computational_score_contribution: Optional[float] = None
    structure_support_contribution: Optional[float] = None
    experimental_activity_contribution: Optional[float] = None
    safety_contribution: Optional[float] = None
    validation_status: str
    has_experimental_data: bool


class ProjectExperimentalValidationSummaryResponse(BaseModel):
    """Project-level summary of experimental validation coverage."""

    model_config = ConfigDict(from_attributes=True)

    project_id: str
    total_candidates: int
    candidates_with_experimental_data: int
    candidates_fully_validated: int
    candidates_failed_validation: int
    candidates_pending: int
    experiment_type_counts: dict[str, int]
    validation_status_counts: dict[str, int]
    measurement_counts: dict[str, int]
    coverage: dict[str, Any]
    top_priority_candidates: list[dict[str, Any]]


class CsvImportError(BaseModel):
    """Single row error from CSV import."""

    row: int
    candidate_id: str
    reason: str


class CsvImportResultResponse(BaseModel):
    """Result of CSV measurement import."""

    model_config = ConfigDict(from_attributes=True)

    project_id: str
    total_rows: int
    success_count: int
    failed_count: int
    created_run_count: int
    created_measurement_count: int
    errors: list[CsvImportError]


# ---------------------------------------------------------------------------
# v0.11-P4: Candidate prioritization
# ---------------------------------------------------------------------------

class PriorityDecisionCreate(BaseModel):
    """Schema for creating a manual priority decision."""

    decision: str = Field(..., max_length=50)
    decision_reason: str = Field(..., min_length=1)
    reviewer: Optional[str] = Field(None, max_length=100)
    notes: Optional[str] = None

    @field_validator("decision")
    @classmethod
    def _validate_decision(cls, v: str) -> str:
        valid = {"SHORTLIST", "HOLD", "REJECT", "NEEDS_REPEAT_EXPERIMENT"}
        if v not in valid:
            raise ValueError(f"decision must be one of {sorted(valid)}")
        return v


class PriorityDecisionResponse(BaseModel):
    """Response after saving a priority decision."""

    model_config = ConfigDict(from_attributes=True)

    candidate_id: str
    decision: dict[str, Any]
    composite_score: Optional[float] = None
    validation_status: str


class CandidatePrioritizationItem(BaseModel):
    """Single candidate in the prioritization list."""

    model_config = ConfigDict(from_attributes=True)

    candidate_id: str
    sequence: str
    composite_score: Optional[float] = None
    experimental_priority_score: Optional[float] = None
    priority_status: str
    validation_status: str
    computational_summary: dict[str, Any]
    experimental_summary: dict[str, Any]
    run_count: int
    measurement_count: int
    decision: Optional[dict[str, Any]] = None


class ProjectCandidatePrioritizationResponse(BaseModel):
    """Full project prioritization response."""

    model_config = ConfigDict(from_attributes=True)

    project_id: str
    summary: dict[str, int]
    candidates: list[CandidatePrioritizationItem]


# ---------------------------------------------------------------------------
# v0.11-P5: Wet-lab validation report export
# ---------------------------------------------------------------------------

class WetlabValidationReportResponse(BaseModel):
    """Response wrapper for wet-lab validation report export (JSON format)."""

    model_config = ConfigDict(from_attributes=True)

    project_id: str
    report_type: str
    generated_at: str
    scientific_boundary: dict[str, bool]
    summary: dict[str, Any]
    candidates: list[dict[str, Any]]

File: src/types/experimentalValidation.ts
/**
 * STAMP Platform v0.11-P2 — Experimental Validation Types
 * Mirrors backend/app/schemas/experimental_validation.py
 */

export type ExperimentType =
  | "MIC"
  | "MBC"
  | "HEMOLYSIS"
  | "CYTOTOXICITY"
  | "SERUM_STABILITY"
  | "PROTEASE_STABILITY"
  | "SALT_STABILITY"
  | "BIOFILM"
  | "RESISTANCE_INDUCTION"
  | "OTHER";

export type RunStatus = "PLANNED" | "RUNNING" | "COMPLETED" | "FAILED" | "INVALIDATED";

export type ValidationStatus =
  | "NOT_EXPERIMENTALLY_VALIDATED"
  | "EXPERIMENT_PLANNED"
  | "PARTIALLY_VALIDATED"
  | "EXPERIMENTALLY_VALIDATED"
  | "VALIDATION_FAILED";

export type MetricName =
  | "MIC_ug_ml"
  | "MBC_ug_ml"
  | "hemolysis_percent"
  | "HC50_ug_ml"
  | "IC50_ug_ml"
  | "cell_viability_percent"
  | "serum_half_life_min"
  | "protease_remaining_percent"
  | "biofilm_inhibition_percent";

export type QualityFlag = "PASS" | "WARNING" | "FAILED" | "NEEDS_REVIEW";

export interface ExperimentalValidationRun {
  id: string;
  project_id: string;
  candidate_id: string;
  experiment_type: ExperimentType;
  organism: string | null;
  strain: string | null;
  protocol_name: string | null;
  protocol_version: string | null;
  operator: string | null;
  experiment_date: string | null;
  status: RunStatus;
  validation_status: ValidationStatus;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface ExperimentalMeasurement {
  id: string;
  validation_run_id: string;
  candidate_id: string;
  metric_name: MetricName;
  value: number | null;
  unit: string | null;
  condition_json: Record<string, unknown>;
  replicate_id: string | null;
  raw_data_path: string | null;
  quality_flag: QualityFlag;
  created_at: string;
  updated_at: string;
}

export interface ExperimentalValidationSummary {
  candidate_id: string;
  overall_validation_status: ValidationStatus;
  run_count: number;
  completed_run_count: number;
  measurements: ExperimentalMeasurement[];
}

export interface CandidateExperimentalPriority {
  candidate_id: string;
  composite_score: number | null;
  experimental_priority_score: number | null;
  priority_status: string;
  computational_score_contribution: number | null;
  structure_support_contribution: number | null;
  experimental_activity_contribution: number | null;
  safety_contribution: number | null;
  validation_status: ValidationStatus;
  has_experimental_data: boolean;
}

export interface ValidationRunCreatePayload {
  project_id: string;
  candidate_id: string;
  experiment_type: ExperimentType;
  organism?: string;
  strain?: string;
  protocol_name?: string;
  protocol_version?: string;
  operator?: string;
  experiment_date?: string;
  notes?: string;
}

export interface MeasurementCreatePayload {
  validation_run_id: string;
  candidate_id: string;
  metric_name: MetricName;
  value?: number;
  unit?: string;
  condition_json?: Record<string, unknown>;
  replicate_id?: string;
  quality_flag?: QualityFlag;
}

export interface ProjectExperimentalValidationSummary {
  project_id: string;
  total_candidates: number;
  candidates_with_experimental_data: number;
  candidates_fully_validated: number;
  candidates_failed_validation: number;
  candidates_pending: number;
  experiment_type_counts: Record<string, number>;
  validation_status_counts: Record<string, number>;
  measurement_counts: Record<string, number>;
  coverage: {
    candidates_with_any_measurement: number;
    candidates_with_mic: number;
    candidates_with_safety: number;
    experimental_coverage_percent: number;
  };
  top_priority_candidates: Array<{
    candidate_id: string;
    sequence: string;
    experimental_priority_score: number | null;
    priority_status: string;
    validation_status: ValidationStatus;
  }>;
}

export interface CsvImportError {
  row: number;
  candidate_id: string;
  reason: string;
}

export interface CsvImportResult {
  project_id: string;
  total_rows: number;
  success_count: number;
  failed_count: number;
  created_run_count: number;
  created_measurement_count: number;
  errors: CsvImportError[];
}

export interface PriorityDecision {
  decision: 'SHORTLIST' | 'HOLD' | 'REJECT' | 'NEEDS_REPEAT_EXPERIMENT';
  decision_reason: string;
  reviewer: string;
  reviewed_at: string;
  manual_override: boolean;
  notes: string | null;
}

export interface CandidatePrioritizationItem {
  candidate_id: string;
  sequence: string;
  composite_score: number | null;
  experimental_priority_score: number | null;
  priority_status: string;
  validation_status: ValidationStatus;
  computational_summary: {
    pepmlm_source: string | null;
    mean_plddt: number | null;
    pdockq: number | null;
    interaction_energy_kcal_mol: number | null;
  };
  experimental_summary: Record<string, unknown>;
  run_count: number;
  measurement_count: number;
  decision: PriorityDecision | null;
}

export interface ProjectCandidatePrioritization {
  project_id: string;
  summary: {
    candidate_count: number;
    ready_for_review: number;
    needs_more_data: number;
    safety_concern: number;
    validation_failed: number;
    shortlisted: number;
    rejected: number;
  };
  candidates: CandidatePrioritizationItem[];
}

export interface PriorityDecisionPayload {
  decision: 'SHORTLIST' | 'HOLD' | 'REJECT' | 'NEEDS_REPEAT_EXPERIMENT';
  decision_reason: string;
  reviewer?: string;
  notes?: string;
}

// v0.11-P5: Wet-lab validation report export
export interface WetlabValidationReport {
  project_id: string;
  report_type: string;
  generated_at: string;
  scientific_boundary: Record<string, boolean>;
  summary: {
    candidate_count: number;
    candidates_with_measurements: number;
    experimental_coverage_percent: number;
    shortlisted: number;
    rejected: number;
    needs_more_data: number;
    safety_concern: number;
    ready_for_review: number;
    validation_failed: number;
  };
  candidates: Array<{
    candidate_id: string;
    sequence: string;
    validation_status: ValidationStatus;
    experimental_priority_score: number | null;
    priority_status: string;
    manual_decision: {
      decision: string;
      decision_reason: string;
      reviewer: string;
      reviewed_at: string;
    } | null;
    experimental_measurements: Array<{
      experiment_type: string;
      metric_name: string;
      value: number;
      unit: string;
      quality_flag: string;
      replicate_id: string | null;
    }>;
    computational_reference: {
      composite_score: number | null;
      mean_plddt: number | null;
      pdockq: number | null;
      interaction_energy_kcal_mol: number | null;
    };
    run_count: number;
    measurement_count: number;
  }>;
}
